Метод $k$-ближайших соседей

Прочитаем данные.

In [1]:
import pandas as pd
train_data = pd.read_csv("KNN.csv", delimiter=',', index_col='id')

Вывод всех строк данных

In [2]:
train_data

X   Y  Class
id               
1   42  53      0
2   47  50      0
3   46  69      1
4   33  67      0
5   10  37      0
6   40  22      0
7   76  37      0
8   81  96      1
9   32  90      0
10  82  70      0

Отберем данные для предикторов и удалим столбец Class

In [3]:
X = pd.DataFrame(train_data.drop(['Class'], axis=1))

Отберем столбец Class для отклика

In [4]:
y = pd.DataFrame(train_data['Class']).values.ravel()

Подключим библиотеку и зададим параметры модели:
n_neighbors — число соседей;
p — используемое расстояние. Манхэттенское расстояние — p=1, евклидово расстояние — p=2.

Посчитаем по евклидову расстоянию.

In [5]:
from sklearn.neighbors import KNeighborsClassifier
neigh = KNeighborsClassifier(n_neighbors=3, p=2)
neigh.fit(X, y)

KNeighborsClassifier(n_neighbors=3)

Новый объект, который необходимо классифицировать: Х=67, У=95

In [6]:
NewObject = [67, 95]

Определим назначенный класс по евклидлвву расстоянию:


In [7]:
neigh.predict([NewObject])

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


array([1])

Получили класс 1.

Вероятности отнесения к классам 0 и 1 соответственно:

In [8]:
neigh.predict_proba([NewObject])

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


array([[0.33333333, 0.66666667]])

Вероятность отнесения к 1 классу - 67%

Вывод расстояний до $k$ соседей в порядке удаления (первый массив — расстояния, второй — идентификаторы объектов).

In [9]:
neigh.kneighbors([NewObject])

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


(array([[14.03566885, 29.15475947, 33.42154993]]), array([[7, 9, 2]]))

Получены расстояния до ближайщих сосебей: 14.03566885, 29.15475947, 33.42154993.
Их индексы [7, 9, 2], значит id = [8, 10, 3]

Посчитаем по манхэттонскому расстоянию

In [10]:
neigh = KNeighborsClassifier(n_neighbors=3, p=1)
neigh.fit(X, y)

KNeighborsClassifier(n_neighbors=3, p=1)

Новый объект, который необходимо классифицировать: Х=67, У=95

In [11]:
NewObject = [67, 95]

Определим назначенный класс по манхэттонскому расстоянию:

In [12]:
neigh.predict([NewObject])

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


array([0])

Получили класс 0.

Вероятности отнесения к классам 0 и 1 соответственно:

In [13]:
neigh.predict_proba([NewObject])

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


array([[0.66666667, 0.33333333]])

Вероятность отнесения к классу 0 - 67%

Вывод расстояний до  k  соседей в порядке удаления (первый массив — расстояния, второй — идентификаторы объектов).

In [14]:
neigh.kneighbors([NewObject])

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  warnings.warn(


(array([[15., 40., 40.]]), array([[7, 9, 8]]))

Получены расстояния до ближайщих сосебей: 15, 40, 40. Их индексы [7, 9, 8], значит id = [8, 10, 9]

В зависимости от способа расчета расстояния один и тот же объект может быть отнесен в разные группы.